### PR TITLE
fix: forward push received app in foreground events to 3rd party callbacks

### PR DESCRIPTION
**Pull request by levibostian**
Part of: https://linear.app/customerio/issue/MBL-138/[bug]-ios-customers-receive-callbacks-from-3rd-party-sdks-for-pushes

When a push is received while app in foreground, send the push event to other push event handlers in host app. Giving customers the opportunity to process the push event. Previously, only non-CIO pushes would be sent to other push event handlers.

Notes:

* CIO SDK waits for all push handlers to call completion handler for push processing. Not calling completion handler may result in push not being shown while app is in foreground.

Push event handlers must call async completion handler for CIO SDK to handle push events. Test function `test_givenMultiplePushHandlers_givenCioPushDelivered_givenOtherPushHandlerDoesNotCallCompletionHandler_expectCompletionHandlerDoesNotGetCalled` reproduces scenario.

Waiting for completion handler prevents premature termination of push processing logic by iOS. Apple documentation states completion handler must be called after processing push event. https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/usernotificationcenter(_:willpresent:withcompletionhandler:)

* The SDK now determines if a push should be shown for non-CIO pushes and fixes bug in `PushEventHandlerProxy`.

This change had to be made as part of waiting for all push handlers to complete async operation.

Behavior before: The push handler that called the async completion handler first would be the one that decides if the push is displayed on the device.

After behavior: The SDK uses Apple's default of not displaying push. The SDK will display the push if at least 1 of the push handlers decides to display push.

This aligns with Apple's default behavior of not displaying a push unless at least one push handler decides to display it.https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/usernotificationcenter(_:willpresent:withcompletionhandler:)

---

**Stack**:
- #634
- #633
- #601
- #590
- #584 ⬅
- #583

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift [hunk 1]
```diff
@@ -61,8 +61,29 @@ class PushEventHandlerProxyImpl: PushEventHandlerProxy {
...
+                if delegateShouldDisplayPushResult {
+                    shouldDisplayPush = true
+                }
+
+                numberOfAsyncOperationsCompleted += 1
+
+                // Do not call the function's completionHandler until *all* delegates have had their completion handler called.
+                if numberOfAsyncOperationsExpected == numberOfAsyncOperationsCompleted {
+                    completionHandler(shouldDisplayPush)
+                }
```
Comment (Ahmed-CIO) [inline Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: There are race conditions in this code that can result in the completionHandler in the if conditions runs too early or never run

**Review by Ahmed-CIO — CHANGES_REQUESTED** [review]

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift (hunk 1 above)
Comment (levibostian) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: "never run" - I suggest we cover that use case in [this thread](https://github.com/customerio/customerio-ios/pull/583#discussion_r1509511361). 

"runs too early" - Could you give steps on how to reproduce this?

Comment (Ahmed-CIO) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: > "never run" - I suggest we cover that use case in https://github.com/customerio/customerio-ios/pull/583#discussion_r1509511361.

This has nothing to with the mentioned thread. This could happen if the race condition caused the counter to miss one count. Then it will never reach the `numberOfAsyncOperationsExpected`.

> "runs too early" - Could you give steps on how to reproduce this?

Hard to explain without lengthy explanation but let's try.
in a simple scenario where `numberOfAsyncOperationsExpected` is set to 2.
Now imagine two threads increase the counter; the first one to go past line 80 after the two threads increased the counter, will meet the condition and call the completion handler, in undetermined order. Then the second one, will call it again.

Basically, there is a race condition, all sort of shenanigans can go.

What's even worse with such simple value (int in this case) that it can't be torn apart (there are exceptions but not needed here); meaning you won't access it in a way that ends up accessing half-baked memory or getting seg_fault errors. So you will never see a crash out of that, but you would see random behavior.

**Review by Ahmed-CIO — CHANGES_REQUESTED** [review]

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift (hunk 1 above)
Comment (Shahroz16) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: I agree with the race condition possibility here, this is why i suggested using `Dispatch Group` in my comment in the other [PR](https://github.com/customerio/customerio-ios/pull/583/files#r1510386994).

**Review by Shahroz16 — COMMENTED** [review]
similar comments and concerns as to the other [PR](https://github.com/customerio/customerio-ios/pull/583).

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift (hunk 1 above)
Comment (levibostian) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: I am on the same page with what you mean by race condition now. Thanks. 

Going off of Shahroz's suggestion, DispatchGroup could be useful here. I don't currently agree with implementing a timeout, although I do understand the value of it. I think the OS should enforce the timeout. It's important to me that we keep our code as close to OS behavior as possible for customers. 

With the timeout code removed, a DisptchGroup code snippet could look similar to this: 

```swift  
let dispatchGroup = DispatchGroup()

for (_, delegate) in nestedDelegates {
    dispatchGroup.enter()

    delegate.onPushAction(pushAction) {
        dispatchGroup.leave()
    }
}

dispatchGroup.notify(queue: .main) {
    completionHandler()
}
```

Unless I hear differently, I will update both PRs with this pattern. I will do that [after working on this ticket](https://linear.app/customerio/issue/MBL-154/process-push-event-before-we-call-3rd-party-callback-functions) to give the team more time to review this PR and give suggestions.

Comment (Shahroz16) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: Are we okay with the fact, we won't be opening deeplinks for a user if any of the `completionHandler()` is not called from any of the delegates.

Comment (Shahroz16) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: Are we okay with the fact, we won't be opening deeplinks for a user if any of the `completionHandler()` is not called from any of the delegates?

Comment (Shahroz16) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: additionally, i think @Ahmed-CIO mentioned they would not prefer to go with `DispatchGroup`, so I would wait for him to raise any concern here before we proceed with it.

Comment (Ahmed-CIO) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: Thanks both for the farther elaboration here.
The point about deep linking is also really important one. Not opening deep linking if a delegate didn't call the completionHandler can also fire back at us.

I will read through the updates I missed and come back to this one again

Comment (Ahmed-CIO) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: Alright, after quick catch, I also did quick refresher about DispatchGroup and my worry about causing deadlock isn't actually valid. 
The only thing that still valid is using dispatchGroup.enter() must be balanced with dispatchGroup.leave(). While in the above code this is the case and easy to reason about, this is prone to all sort of issues one used to do in the early days of manually calling alloc/release. However, this is not as a hard blocker as the point I initially raised and turns out to be wrong.
My recommendation: if it is not too much work to create a small actor that only contains the parts that needs to synced, then let's do that as it is likely to be much cleaner solution. Otherwise, go with DispatchGroup. If it helps, here is a quick non-fully-validated example of how actor could help
```swift
actor PushDelegateAsyncOperationsTracker {
    private var numberOfAsyncOperationsExpected: Int
    private var shouldShowNotification: Bool = false
    private let completionHandler: (Bool) -> Void
    
    init(_ numberOfAsyncOperationsExpected: Int,
         completionHandler: @escaping (Bool) -> Void) {
        self.numberOfAsyncOperationsExpected = numberOfAsyncOperationsExpected
        self.completionHandler = completionHandler
    }
    
    
    func onAsyncOperationExecuted(_ shouldShowNotification: Bool) {
        self.numberOfAsyncOperationsExpected -= 1
        self.shouldShowNotification = self.shouldShowNotification || shouldShowNotification
        
        if self.numberOfAsyncOperationsExpected == 0 {
            self.completionHandler(self.shouldShowNotification)
        }
    }
}

delegate.shouldDisplayPushAppInForeground(push)  { delegateShouldDisplayPushResult in
        Task {
            await operationsTracker.onAsyncOperationExecuted(delegateShouldDisplayPushResult)
        }
 })
    
```

Comment (levibostian) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: Wrote some code and shared it with team in Slack. After response, I'll update this PR.

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift [hunk 2]
```diff
@@ -61,8 +61,28 @@ class PushEventHandlerProxyImpl: PushEventHandlerProxy {
...
+        Task {
+            // 2+ other push event handlers may exist in app. We need to decide if a push should be displayed or not, by combining all the results from all other push handlers.
+            // To do that, we start with Apple's default value of: do not display.
+            // If any of the handlers return result indicating push should be displayed, we return true.
+            // Apple docs: https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/usernotificationcenter(_:willpresent:withcompletionhandler:)
+            var shouldDisplayPush = false
+
+            // Wait for all other push event handlers to finish before calling the completion handler.
+            // Each iteration of the loop waits for the push event to be processed by the delegate.
+            for delegate in nestedDelegates.values {
```
Comment (Shahroz16) [inline Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:73]: same comment about `nestedDelegates` as in the other [PR](https://github.com/customerio/customerio-ios/pull/583#discussion_r1516079941).

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift [hunk 3]
```diff
@@ -61,8 +61,28 @@ class PushEventHandlerProxyImpl: PushEventHandlerProxy {
...
+                        if delegateShouldDisplayPushResult {
+                            shouldDisplayPush = true
+                        }
+
+                        continuation.resume()
+                    })
+                }
+            }
+            // After the loop finishes, call the completion handler to indicate the event has been fully processed by all delegates.
+            completionHandler(shouldDisplayPush)
```
Comment (Shahroz16) [inline Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: an existing conversation going on it, regarding we may not show the push in forground if the delegates didn't give callback.

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift (hunk 2 above)
Comment (levibostian) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:73]: This has been updated and ready for re-review.

File: Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift (hunk 3 above)
Comment (levibostian) [reply Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift:85]: This problem has been resolved with new PR: https://github.com/customerio/customerio-ios/pull/601

**Review by Ahmed-CIO — APPROVED** [review]
Check the comments in case you want to address the nit comment

**Review by Shahroz16 — APPROVED** [review]

**Review by Ahmed-CIO — APPROVED** [review]

**Comment by levibostian** [comment]
✓ Commit merged in pull request [#634](https://github.com/customerio/customerio-ios/pull/634)
